DancebotESP32/src/DemobotNetwork.cpp

```diff
@@ -0,0 +1,242 @@
+    /* If not connected to a network or doesn't have an ip address. */
+    // TODO: don't know if !_ipaddress will work
+    if (!_connected || !_ipaddress) return -1;
+
+    String queryPath = "http://" + IpAddress2String(_ipaddress) + endpoint;
+    queryPath += '?';
+    
+    /* Append keys and values to the path. */
+    for (int i = 0; i < argSize; i++) {
+        queryPath += keys[i];
+        queryPath += '=';
+        queryPath += vals[i];
+        if (i < (argSize - 1)) {
+            queryPath += '&';
+        }
+    }
+
+    /* Send the request. */
+    http->begin(queryPath.c_str());
+    int responseCode = http->GET();
+    if (responseCode > 0) {
+        String str = String(http->getString());
+        response = &(str);
+    }
+    return responseCode;
+}
+
+int DemobotNetwork::sendPOSTRequest(const String endpoint, const String keys[], const String vals[], int argSize, String *response) const {
+    /* If not connected to a network or doesn't have an ip address. */
+    // TODO: don't know if !_ipaddress will work
+    if (!_connected || !_ipaddress) return -1;
+
+    String queryPath = "http://" + IpAddress2String(_ipaddress) + endpoint;
+    http->begin(queryPath);
+    http->addHeader("Content-Type", "application/x-www-form-urlencoded");
+    String data = "";
+
+    /* Append keys and values to the path. */
+    for (int i = 0; i < argSize; i++) {
+        data += keys[i];
+        data += '=';
+        data += vals[i];
+        if (i < (argSize - 1)) {
+            data += '&';
+        }
+    }
+
+    /* Send the request. */
+    int responseCode = http->POST(data);
+    if (responseCode > 0) {
+        String str = String(http->getString());
+        response = &(str);
+    }
+    return responseCode;
+}
+
+char* DemobotNetwork::getNetworkSSID() const {
+    return _SSID;
+}
+
+char* DemobotNetwork::getNetworkPassword() const {
+    return _PASSWORD;
+}
+
+void DemobotNetwork::shutdownNetwork() {
+    delete[] credentialsLog;
+    delete http;
+}
+
+IPAddress DemobotNetwork::getIPAddress() const {
+    return _ipaddress;
+}
+
+String DemobotNetwork::IpAddress2String(const IPAddress ipAddress) const {
+    return String(ipAddress[0]) + String(".") +\
+        String(ipAddress[1]) + String(".") +\
+        String(ipAddress[2]) + String(".") +\
+        String(ipAddress[3]) ;
+}
+
+/** Private methods. */
+
+int DemobotNetwork::hash(char* valptr) const {
+    /**
+     * This is dumb hash and really shouldn't be used for anything but small
+     * subsets of problems where we're pretty confident that collisions are small.
+     * 
+     * We're just taking the char and multiplying it by 2 by its place:
+     * (i.e. hello -> h*2^0 + e*2^1 + l*2^2 + l*2^3 + o*2^4).
+     * Preferably, for no collisions, you'd want to have everything be a power
+     * of 62 (52 characters - lower and upper - and 10 digits), i.e. 36^place. But this is
```

Comment:
Should be 62^place if I know how permutations work @amadraccoon101 guess I'm gonna fail 351k again